### PR TITLE
[VCDA-4679] Attempt adding network to the VM if it is missing

### PR DESCRIPTION
**Pull request by Anirudh9794**
Signed-off-by: Aniruddha Shamasundar <aniruddha.9794@gmail.com>

## Description
Please provide a brief description of the changes proposed in this Pull Request

- make attempt to add network to the VM is network connection section for the VM is missing.
- Tested by creating multiple node pools, resizing up one node pool and resizing down one node pool simultaneously.

## Checklist
- [x] tested locally
- [ ] updated any relevant dependencies
- [ ] updated any relevant documentation or examples

## API Changes
Are there API changes?
- [ ] Yes
- [x] No

If yes, please fill in the below

1. Updated conversions?
    - [ ] Yes
    - [ ] No
    - [x] N/A
2. Updated CRDs?
    - [ ] Yes
    - [ ] No
    - [x] N/A
3. Updated infrastructure-components.yaml?
    - [ ] Yes
    - [ ] No
    - [x] N/A
4. Updated `./examples/capi-quickstart.yaml`?
    - [ ] Yes
    - [ ] No
    - [x] N/A
5. Updated necessary files under `./infrastructure-vcd/v1.0.0/`?
   - [ ] Yes
   - [ ] No
   - [x] N/A

## Issue
If applicable, please reference the relevant issue

Fixes #

- - -
This change is [Reviewable](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/310)

### CODE REVIEW

**Review by sahithi — APPROVED** [review]
:lgtm: after addressing the comments

> *[Reviewable](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/310#-:-NEqEX810HH9iWWm8a-d:b-3tyihq)* status: 0 of 1 files reviewed, 1 unresolved discussion (waiting on @Anirudh9794)

___
*[`controllers/vcdmachine_controller.go` line 564 at r1](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/310#-NEqEHyu4ZWrwgZmoQps:-NEqEHyu4ZWrwgZmoQpt:bpyv9zp) ([raw file](https://github.com/vmware/cluster-api-provider-cloud-director/blob/ba7298a36ae46498cfaf5f740b746f0fc9971f2a/controllers/vcdmachine_controller.go#L564)):*
> ```Go
> 		err := vm.UpdateNetworkConnectionSection(networkConnectionSection)
> 		if err != nil {
> 			klog.Warningf("Failed to update VM [%s] with network connection section: [%v]", vm.VM.Name, err)
> ```

klog can be replaced by log.v(4)

**Review by Anirudh9794 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/310)* status: 0 of 1 files reviewed, 1 unresolved discussion (waiting on @sahithi)

___
*[`controllers/vcdmachine_controller.go` line 564 at r1](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/310#-NEqEHyu4ZWrwgZmoQps:-NEvOasB9QWWrlr1igKz:b-896fix) ([raw file](https://github.com/vmware/cluster-api-provider-cloud-director/blob/ba7298a36ae46498cfaf5f740b746f0fc9971f2a/controllers/vcdmachine_controller.go#L564)):*
<details><summary><i>Previously, sahithi (Sahithi Ayloo) wrote…</i></summary><blockquote>

klog can be replaced by log.v(4)
</blockquote></details>

Done.

**Review by arunmk — APPROVED** [review]
:lgtm: with some small nits

> Reviewed 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/310#-:-NEwJhwj8JCPKa6w65fU:bxjmvxw)* status: all files reviewed, 3 unresolved discussions (waiting on @Anirudh9794 and @sahithi)

___
*[`controllers/vcdmachine_controller.go` line 571 at r2](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/310#-NEwJUiL5EGqyQ-voU1C:-NEwJUiL5EGqyQ-voU1D:b-th3vjc) ([raw file](https://github.com/vmware/cluster-api-provider-cloud-director/blob/e1ae0fd41bbbe5fdcfb1338c4cd2875853577f0d/controllers/vcdmachine_controller.go#L571)):*
> ```Go
> 
> 	if vm.VM.NetworkConnectionSection.NetworkConnection[0] == nil {
> 		log.V(4).Info("Attempting to update the VM [%s] with network connection section: [%#v]", vm.VM.Name, networkConnectionSection)
> ```

Can we not merge this if also with the previous one itself? it is just another `||` I think. The short-circuit should ensure that the zeroth element is present.
___
*[`controllers/vcdmachine_controller.go` line 582 at r2](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/310#-NEwJfZy-K6_-_fBWF5E:-NEwJfZy-K6_-_fBWF5F:b3aomdf) ([raw file](https://github.com/vmware/cluster-api-provider-cloud-director/blob/e1ae0fd41bbbe5fdcfb1338c4cd2875853577f0d/controllers/vcdmachine_controller.go#L582)):*
> ```Go
> 
> 	if vm.VM.NetworkConnectionSection.NetworkConnection[0].IPAddress == "" {
> 		log.Info("Attempting to update the VM [%s] with network connection section: [%#v]", vm.VM.Name, networkConnectionSection)
> ```

Likewise this one also

**Review by Anirudh9794 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/310)* status: 0 of 1 files reviewed, 3 unresolved discussions (waiting on @arunmk and @sahithi)

___
*[`controllers/vcdmachine_controller.go` line 571 at r2](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/310#-NEwJUiL5EGqyQ-voU1C:-NEwPgC5ExSUiv83lHxh:b-896fix) ([raw file](https://github.com/vmware/cluster-api-provider-cloud-director/blob/e1ae0fd41bbbe5fdcfb1338c4cd2875853577f0d/controllers/vcdmachine_controller.go#L571)):*
<details><summary><i>Previously, arunmk (Arun M. Krishnakumar) wrote…</i></summary><blockquote>

Can we not merge this if also with the previous one itself? it is just another `||` I think. The short-circuit should ensure that the zeroth element is present.
</blockquote></details>

Done.
___
*[`controllers/vcdmachine_controller.go` line 582 at r2](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/310#-NEwJfZy-K6_-_fBWF5E:-NEwPgY_12-nUKAbNHEq:b-896fix) ([raw file](https://github.com/vmware/cluster-api-provider-cloud-director/blob/e1ae0fd41bbbe5fdcfb1338c4cd2875853577f0d/controllers/vcdmachine_controller.go#L582)):*
<details><summary><i>Previously, arunmk (Arun M. Krishnakumar) wrote…</i></summary><blockquote>

Likewise this one also
</blockquote></details>

Done.

**Pull request merged**
